### PR TITLE
Fix #1273: Add tests for hints and solutions

### LINKED ISSUE #1273: Introduce UI tests for hints & solutions
Hints & solutions logic was updated in #500, but an unanticipated regression delayed its merge by a week.
This regression could have been caught earlier if robust UI tests were in place. Currently, there is no comprehensive test coverage for hint & solution flows in the app, which risks future regressions going undetected.

## Solution Approach

Introduce **dedicated UI tests** that:

Introduce **dedicated UI tests** that:

1. **Verify hint bulb visibility depending on user actions**
   * `testStateFragment_forHintsAndSolution_incorrectInput_hintBulbContainerIsNotVisible`
   * `testStateFragment_forHintsAndSolution_incorrectInputTwice_hintBulbContainerIsVisible`
   * `testStateFragment_forHintsAndSolution_incorrectInput_thenCorrectInput_doesNotShowHint`
   * `testStateFragment_withNoHints_doesNotShowHintBulb`

2. **Ensure correct accessibility labels for hints & solutions**
   * `testStateFragment_showHintsAndSolutionBulb_dotHasCorrectContentDescription`
   * `testStateFragment_showHintsAndSolutionBulb_bulbHasCorrectContentDescription`

3. **Validate the lifecycle of hints and solutions (timing, showing, dismissal)**
   * `testStateFragment_hintShown_afterDelayAndSubsequentHintChanges_doNotRedrawActiveAnswer`
   * `testStateFragment_revealSolution_afterHints_showsSolutionAndDisablesFurtherHints`
   * `testStateFragment_repeatedHints_doNotStackOrDuplicate`
   * `testStateFragment_hint_withNullText_showsFallbackOrNoCrash`

4. **Test navigation scenarios (moving between questions/explorations)**
   * `testQuestionPlayer_hintFlow_acrossQuestionNavigation_restoresCorrectState`

5. **Cover edge cases like no hints, null hints, or duplicate updates**
   * (Already covered by: `testStateFragment_withNoHints_doesNotShowHintBulb`,  
     `testStateFragment_hint_withNullText_showsFallbackOrNoCrash`,  
     `testStateFragment_repeatedHints_doNotStackOrDuplicate`)

6. **Add sanity Espresso flows to ensure correctness on device-like conditions**
   * `testHintsSolution_basicFlow_espresso_endToEnd`

These tests should live under `StateFragmentTest` or `QuestionPlayerTest` depending on scope, with reusable helpers like `StateViewHolderTypeMatcher` for RecyclerView navigation.

## Acceptance Criteria

* All major hint & solution behaviors are covered by automated tests.
* Tests run successfully .
* Future regressions (like the one in #500) will be reliably caught.
* Coverage ensures proper user experience (no premature hints, correct accessibility, no keyboard disruption).
* Include a screenshot of the passing tests.

### PR DESCRIPTION
**Pull request by BenHenning**
Fixes #1273.

This introduces tests for hints & solutions. Note that these are Robolectric-only tests because a lot of UI screens need to be quickly driven through, and we need to potentially wait long periods of time (so utilizing fake time is really helpful). Consider that locally, these tests seem to run in about ~40 seconds. Some of the tests are verifying scenarios where the user is waiting 60 seconds for a hint to show, and may go through multiple states with questions in order to reach a desired state. These tests heavily rely on the time-related fixes introduced in #1275.

This PR relies on one of the fractions explorations for driving different hints & solutions flows. To keep things simple, it actually drives through the exploration like a user. In the future, we may want to create dedicated explorations for the different scenarios being tested here.

Note that this PR also is the first test that utilizes proper test dependency replacement for UI components (support added in #1275). This is needed in order to switch the production coroutine dispatchers with test ones that can allow coordination with the fake clock. Subsequently, these tests are not compatible with Espresso which is why they aren't living under sharedTest.

Note also that this new test introduces a nice ``StateViewHolderTypeMatcher`` for quickly scrolling between the elements of the state & questions recycler views. We may want to utilize this in future tests.

### CODE REVIEW

**Comment by rt4914** [comment]
> /cc @rt4914 & @nikitamarysolomanpvt (note that new way to test more timing-sensitive UI flows).
> /cc @anandwana001 per our discussion (this test includes stacked dialog handling with Espresso matchers).

@BenHenning Sounds good. Will take a look.

**Comment by anandwana001** [comment]
> /cc @rt4914 & @nikitamarysolomanpvt (note that new way to test more timing-sensitive UI flows).
> /cc @anandwana001 per our discussion (this test includes stacked dialog handling with Espresso matchers).

Awesome, is it like using the fragment instance in the test file and opening the dialog?

**Comment by BenHenning** [comment]
@anandwana001 in my case the issue was interacting with an alert dialog that was being opened by production code on top of a custom dialog fragment (custom dialog fragments seem to work fine, or at least the hints & solution one does). Espresso wasn't able to interact with the alert dialog, so I needed to retrieve the fragment directly. I think this will work when running the test both locally & on an emulator, but I didn't check since the tests in question are specifically meant for Robolectric.

File: app/src/main/java/org/oppia/app/player/state/testing/StateFragmentTestActivity.kt [hunk 1]
```diff
@@ -6,6 +6,8 @@ import android.os.Bundle
 import org.oppia.app.activity.InjectableAppCompatActivity
 import org.oppia.app.model.HelpIndex
```
Comment (veena14cs) [inline app/src/main/java/org/oppia/app/player/state/testing/StateFragmentTestActivity.kt]: Please remove unused imports.

File: app/src/test/java/org/oppia/app/player/state/StateFragmentLocalTest.kt [hunk 2]
```diff
@@ -0,0 +1,1088 @@
...
+  }
+
+  // TODO(#59): Figure out a way to reuse modules instead of needing to re-declare them.
+  @Singleton
+  @Component(modules = [
+    TestDispatcherModule::class, ApplicationModule::class, NetworkModule::class,
+    LoggerModule::class, ContinueModule::class, FractionInputModule::class,
+    ItemSelectionInputModule::class, MultipleChoiceInputModule::class,
+    NumberWithUnitsRuleModule::class, NumericInputRuleModule::class, TextInputRuleModule::class,
+    InteractionsModule::class, GcsResourceModule::class, GlideImageLoaderModule::class,
```
Comment (veena14cs) [inline app/src/test/java/org/oppia/app/player/state/StateFragmentLocalTest.kt]: I am unable to run this test class. It was giving below error
Screenshot 2020-06-10 at 1 44 49 PM

I tried adding  `@DragDropSortInputRules` annotation in InteractionsModule `provideDragAndDropSortInputInteractionClassifier` method. Error was resolved but after running I got below failed testcases.

Screenshot 2020-06-10 at 1 43 24 PM

**Review by veena14cs — CHANGES_REQUESTED** [review]
Testcases are failing. I want to defer it to other reviewers, if they are facing same issue or I am missing something here.

File: app/src/test/java/org/oppia/app/player/state/StateFragmentLocalTest.kt (hunk 2 above)
Comment (BenHenning) [reply app/src/test/java/org/oppia/app/player/state/StateFragmentLocalTest.kt]: I suspect the test may be out of date from ``develop``. I'll need to update it and see if I missed an import from new classes being added.

I'm not sure why the shadow exception is being triggered here.

Comment (BenHenning) [reply app/src/test/java/org/oppia/app/player/state/StateFragmentLocalTest.kt]: I can verify that I hit these issues locally. I think a PR that went in 2 weeks ago triggered all the tests to start breaking. I'm investigating what needs to be done to fix it.

Comment (BenHenning) [reply app/src/test/java/org/oppia/app/player/state/StateFragmentLocalTest.kt]: Okay, I think it's fixed now. PTAL.

File: app/src/main/java/org/oppia/app/player/state/testing/StateFragmentTestActivity.kt (hunk 1 above)
Comment (BenHenning) [reply app/src/main/java/org/oppia/app/player/state/testing/StateFragmentTestActivity.kt]: Nice catch--done!

**Review by veena14cs — APPROVED** [review]
LGTM all testcases are passing.

File: app/src/test/java/org/oppia/app/player/state/StateFragmentLocalTest.kt (hunk 2 above)
Comment (veena14cs) [reply app/src/test/java/org/oppia/app/player/state/StateFragmentLocalTest.kt]: It is fixed. Thanks!

**Comment by BenHenning** [comment]
Thanks @veena14cs!

FYI everyone else: I'm going to go ahead and merge this. Please follow up with me if you have any concerns.

**Comment by BenHenning** [comment]
NB: CircleCI was passing on this branch, see:

Screenshot 2020-06-26 at 11 43 54 AM

I'm going to update with develop & re-run the GitHub actions tests per requirements, but I may force the PR submission given that past CircleCI tests are indeed passing.

**Review by aggarwalpulkit596 — APPROVED** [review]

**Comment by BenHenning** [comment]
Reminder to self to update this with the latest develop.

**Comment by BenHenning** [comment]
This is now up-to-date with the latest develop, and the tests have been updated to pass again with recent hints & solutions changes.
